✨ add e2e tests on the ginkgo framework

effi-ofer opened the pull request:
<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary
This PR re-writes the current e2e test cases (multi-cluster-deployment, update-delete, singleton, filtering) using the ginkgo framework.  

Notes:
1. I am currently leaving the existing test as they are, we may want to use some of them for examples.  
2. The test takes about 10min, almost 8min of which is the setup.

## Related issue(s)

Fixes #1612

pdettori commented:
are issues #1710 and #1711 impacting also the basic e2e test scenarios that were initially implemented with bash, or only the specific tests where placement is deleted or altered ?

effi-ofer commented:
> are issues #1710 and #1711 impacting also the basic e2e test scenarios that were initially implemented with bash, or only the specific tests where placement is deleted or altered ?

I haven't observed these failures in the bash e2e test scenarios.  But note that I can reproduce these issues from the shell outside of ginkgo.

MikeSpreitzer commented:
Are #1710 and #1711 being fixed by the pluggable transport work?

MikeSpreitzer commented on Makefile:
```diff
@@ -173,6 +173,11 @@ ko-build-local: test ## Build local container image with ko
 	docker tag ko.local/${CMD_NAME}:${IMAGE_TAG} ${IMG}
 
+.PHONY: ko-build-local-bare
+ko-build-local-bare: manifests generate ## Build local container image with ko
```
There is a lot of material duplicated here; that is unfortunate. This is the same as `ko-build-local` except that it does not depend on the `test` target, right? I assume that is to avoid a circular dependency among `make` targets? I thought that `make test` was for running _unit_ tests. I think we can avoid circular dependency _and_ avoid duplicating Makefile contents if we distinguish three kinds of tests: unit tests, integration tests, and E2E tests.  See how I did some of that in #1769 .

effi-ofer replied:
As usual, great comment.  I foolishly didn't realize the make test is selective.  I switched to your suggestion.

MikeSpreitzer commented on test/e2e/common/setup-kubestellar.sh:
```diff
@@ -22,6 +22,10 @@ if [[ "$KFLEX_DISABLE_CHATTY" = true ]] ; then
    echo "disable_chatty_status = $disable_chatty_status"
 fi
 
+SRC_DIR="$(cd $(dirname "${BASH_SOURCE[0]}") && pwd)"
+COMMON_SRCS="${SRC_DIR}/../common"
+source "$COMMON_SRCS/setup-shell.sh"
```
Why is this stuff here? I do not see any use of it later in this script (which is not `source`d, so the environment modifications have not effect on anything once this script exits).

effi-ofer replied:
Doesn't the setup-shell.sh contains the wait-for-cmd that setup-kubestellar.sh uses?

MikeSpreitzer replied:
Oh, right, this PR execs this script on its own.

MikeSpreitzer commented on test/util/util.go:
```diff
@@ -0,0 +1,202 @@
+	return config
+}
+
+func CreateClient(config *rest.Config) *kubernetes.Clientset {
```
FYI: I would add something to the name of this func, such as "K8s", indicating which kind of client it is making.

effi-ofer replied:
Done

MikeSpreitzer reviewed (commented):
/lgtm
Thanks!

nirrozenbaum commented:
if pluggable transport fixes the blockers for this issue, I think it’s worth spending more time on getting this to work

effi-ofer commented:
> On second thought, there is something missing here. I do not see updates to instructions. BTW, I do not see removal of the existing run_test.sh bash scripts; I assume that is deliberate. It would be clearer if there were instructions updates. This will be easier after #1817 merges.
> 
> /remove-lgtm

Hey @MikeSpreitzer, I think this PR is ready for review.  I added documentation in the test/e2e directory.  Let me know if I should document elsewhere as well.  I updated the .github workflows but intentionally kept the shell scripts in case we want to use them for examples.  Please also confirm that I haven't broken the way you are using --released when testing releases.

ezrasilvera commented:
@effi-ofer Does this replace both singleton and multi-cluster tests? 
One performance issue is that the above used to run on two different workflows and could run in parallel .

effi-ofer commented:
> @effi-ofer Does this replace both singleton and multi-cluster tests? 
Yes. And all other tests in the e2e subfolder.

> One performance issue is that the above used to run on two different workflows and could run in parallel .
I removed the other workflows as part of this PR.

MikeSpreitzer commented on .github/workflows/pr-test-e2e.yml:
```diff
@@ -47,13 +47,14 @@ jobs:
-          cd test/e2e/multi-cluster-deployment 
-          KFLEX_DISABLE_CHATTY=true ./run-test.sh $TEST_FLAGS
+          cd test/e2e/ginkgo 
+          KFLEX_DISABLE_CHATTY=true KUBECONFIG=~/.kube/config ginkgo -v $TEST_FLAGS
```
It looks strange to set `KUBECONFIG` to the value that has the same meaning as leaving it unset. Why is that there?

effi-ofer replied:
It looks like ginkgo is somehow setting a different default value of KUBECONFIG.  If I don't set the KUBECONFIG, then the setup script fails when it first attempts to access the kind-kubeflex context.  I'll dig further into it.

MikeSpreitzer replied:
Any news on this?

ezrasilvera commented:
/hold

MikeSpreitzer commented:
I see that this PR **DOES NOT** delete the bash versions of the E2E tests. This is IMPORTANT and GOOD. The bash versions are the only place where we show contributors bash commands for exercising _their version_ of the system.

waltforme commented on test/e2e/ginkgo/multiple_cluster_deployment_test.go:
```diff
@@ -0,0 +1,178 @@
+	})
+
+	ginkgo.Context("multiple WECs", ginkgo.Ordered, func() {
+		ginkgo.It("propegates deployment to the WECs", func() {
```
Typo in 'propegates'?

effi-ofer replied:
Thanks. Fixed.

MikeSpreitzer commented on test/e2e/ginkgo/README.md:
```diff
@@ -0,0 +1,25 @@
+1. Changing the bindingpolicy objectSelector to no longer match should delete the object from the WECs.
+1. Changing the bindingpolicy objectSelector to match should create the object on the WECs.
+1. Delete of the bindingpolicy object should delete the object on the WECs.
+1. Delete of the overlapping bindingpolicy object should not delete the object on the WECs.
```
What overlapping BindingPolicy? This was not mentioned before.

effi-ofer replied:
This previously existed in the update-delete.sh test.  By overlapping bindingpolicy it is meant a binding policy that results in the same objects downsynced into the same clusters.

ezrasilvera replied:
@MikeSpreitzer this is a special case of 2 BPs on the same WDS that update the same object going to the same WEC.  In OCM one of them will "win"  but when we remove one of the BP the object will move to be "owned" by the other one and shouldn't be deleted.

MikeSpreitzer replied:
@ezrasilvera : as @effi-ofer mentioned, there is no "win" here, it is a matter of redundancy.

@effi-ofer: My point here is about the language in this README. Where it refers to "the overlapping bindingpolicy", an English reader expects to have already been told about an "overlapping bindinbpolicy". OTOH, if this text said "an overlapping bindingpolicy" then that expectation would not be there.

ezrasilvera replied:
@MikeSpreitzer It's not redundancy it is a  win in term of ownership.  As a result from the overlapping BP two manifestWorks are created. However on the WEC the OCM agent will treat only one of them as the source for that object so both manifestwork will fight on the ownership.   Once one of the binding is removed --> one of the manifestWork is removed --> the OCM agent will change the ownership to the other manifestwork
In our case it's not so interesting as we are on the same WDS --> same object --> it doesn't matter who wins. But if we do the same test with different WDS and create the same object (name, NS, etc..)  with couple of different fields then the question of "winning ownership" has more meaning.

MikeSpreitzer replied:
@ezrasilvera : sorry, I misread your statement, did not notice the "in OCM" qualification. However, it is not a matter of one BindingPolicy or another winning, the OCM question is about one ManifestWork or another winning. I think that the KubeStellar user does not care about this question, since overlapping BindingPolicies in one WDS will produce redundant content in the ManifestWorks, so it will not matter which one OCM marks as object owner in the WEC.

effi-ofer replied:
I am sorry but I am getting lost with all the explanations :( Is the problem the use of "the" rather than "an"?  If so I agree, and will change.

MikeSpreitzer commented on test/e2e/ginkgo/README.md:
```diff
@@ -0,0 +1,25 @@
+1. Delete of the workload object on WDS deletes the relevant objects on the WECs.
+1. Delete of a bindingpolicy deletes the relevant objects on the WECs.
+1. Singleton status update
+1. Object field filtering for services
```
I think we usually describe this as "cleaning".

What about Job objects?

effi-ofer replied:
You are right.  I think the name filtering is more appropriate, but I'll switch it back to cleaning.

MikeSpreitzer replied:
Filtering is not bad, but "field filtering" is too specific. Cleaning involves removing labels too, for example. "content filtering" would work.

MikeSpreitzer commented on test/e2e/ginkgo/README.md:
```diff
@@ -0,0 +1,25 @@
+
+To test the latest release image, issue `ginkgo -v -- -released=on`
+
+To test a specific test unit use ginkgo's --focus parameter.  For example, `ginkgo -v --focus "survives IMBS vcluster coming down"`
```
These are not "unit" tests. These are end-to-end tests.

effi-ofer replied:
These are not unit tests.  These are test units, i.e. a single operation or set of operations that compose one test within a a test file. Is this term, test unit, no longer popular in the field?  If so I'll stop using it.

MikeSpreitzer replied:
Sorry, I misread. It is OK.

MikeSpreitzer commented on test/e2e/ginkgo/ginkgo_suite_test.go:
```diff
@@ -0,0 +1,74 @@
+	releasedFlag string
+)
+
+func init() {
```
Why not `TestMain` ? (see https://pkg.go.dev/testing@go1.22.1#hdr-Main)

effi-ofer replied:
I have been following the ginkgo recommendation https://onsi.github.io/ginkgo/#supporting-custom-configuration-custom-command-line-flags.  It seems Ginkgo's main contributor doesn't like TestMain https://github.com/onsi/ginkgo/issues/1233#issuecomment-1625913301.

MikeSpreitzer replied:
Hmm, now I do not remember why I suggested that. I do not see the standard "testing" package being used here.

MikeSpreitzer commented on test/e2e/ginkgo/ginkgo_suite_test.go:
```diff
@@ -0,0 +1,74 @@
+
+func init() {
+	flag.StringVar(&releasedFlag, "released", "", "released controls whether we use a release image")
+}
```
How does a user control the debug log verbosity level (the thing that is normally set with `-v=N`)?

effi-ofer replied:
I haven't found a way to pass a log verbosity level with ginkgo.  There is the '-v' flag for additional output, the '-vv' for maximum output, and the '-trace' to have a trace at each failure.  But I haven't found a logging level.  Perhaps we need to set that up as an external flag similar to -released?

MikeSpreitzer replied:
To be sure, this is not a blocker issue. The bash-based tests have fixed verbosity. As long as the verbosity is reasonably high --- high enough to debug problems --- that is good enough.

MikeSpreitzer commented on test/e2e/ginkgo/ginkgo_suite_test.go:
```diff
@@ -0,0 +1,74 @@
+var (
+	ctx          context.Context
+	coreCluster  *kubernetes.Clientset
+	wds          *kubernetes.Clientset
+	ksWds        *ksClient.Clientset
```
I would make these have names with parallel structure. For example, `k8sWds` and `ksWds` or `wdsK8s` and `wdsKS`. And follow through with the other variables here.

effi-ofer replied:
If you don't mind Mike, I'll leave the names as their are for now so we can proceed with the merge.  If you find the naming a pain, ping me privately and I'll change them in a separate PR.

MikeSpreitzer replied:
By "FYI" I mean this is just my opinion on a matter of style, not something that has to be changed in this PR or ever.

MikeSpreitzer commented on test/e2e/ginkgo/ginkgo_suite_test.go:
```diff
@@ -0,0 +1,74 @@
+	ksWds        *ksClient.Clientset
+	wec1         *kubernetes.Clientset
+	wec2         *kubernetes.Clientset
+	ocmWorkImbs  *ocmWorkClient.Clientset
```
We are going with "ITS" rather than "IMBS". Yes, some other stuff is outdated too.

FYI: I would not mind seeing only "work" here rather tan "ocmWork".

effi-ofer replied:
Oh I didn't know about ITS.  I humbly suggest change in name is done all at once, i.e. search and replace all IMBS into ITS.

MikeSpreitzer replied:
We already have a mixture of ITS and IMBS. I think new work should use the new name; cleaning up outdated stuff can be done separately.

effi-ofer replied:
Ok.  I scrubbed the term IMBS and replaced it with ITS.

MikeSpreitzer commented on test/e2e/ginkgo/multiple_cluster_deployment_test.go:
```diff
@@ -0,0 +1,178 @@
+		})
+	})
+
+	ginkgo.Context("object field filtering", func() {
```
I think we more commonly call this "object cleaning".

effi-ofer replied:
I changed it to cleaning.

MikeSpreitzer replied:
https://github.com/effi-ofer/edge-mc/blob/73515206ca859c31aebbf5d56b010dc5e1cfdea7/test/e2e/ginkgo/multiple_cluster_deployment_test.go#L133 still speaks of "field filtering".

effi-ofer replied:
I just scrabbed the word filtering out.

MikeSpreitzer replied:
"Field cleaning" is still too specific. An individual label or annotation is not a "field". Just plain "cleaning" works.

effi-ofer replied:
The word field is removed.

MikeSpreitzer commented on test/e2e/ginkgo/multiple_cluster_deployment_test.go:
```diff
@@ -0,0 +1,178 @@
+	})
+
+	ginkgo.Context("object field filtering", func() {
+		ginkgo.It("properly starts a service", func() {
```
We should add a test for Job cleaning, since that is implemented too. Does not have to be done in this PR.

effi-ofer replied:
https://github.com/kubestellar/kubestellar/issues/1867

MikeSpreitzer commented on test/util/util.go:
```diff
@@ -0,0 +1,346 @@
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
```
Many of the funcs in this file are not general utilities, they are only for use in gingko tests. Maybe have a `testutil` package? If not, at least name this file something more specific. "util/util" says nothing.

effi-ofer replied:
It's not util/util.  It's test/util/util.go.  Or test/util when imported.  It is located under test and is included as "github.com/kubestellar/kubestellar/test/util".  Isn't changing util to testutil redundant?  You'll end up with "github.com/kubestellar/kubestellar/test/testutil".  If you think this is important, I'll make the change.

MikeSpreitzer replied:
Sorry, I overlooked the "test" part.

MikeSpreitzer commented on test/util/util.go:
```diff
@@ -0,0 +1,346 @@
+			},
+		},
+	}
+	wds.ControlV1alpha1().BindingPolicies().Create(ctx, &bindingPolicy, metav1.CreateOptions{})
```
I think error checking here would be helpful

effi-ofer replied:
done

MikeSpreitzer commented on test/util/util.go:
```diff
@@ -0,0 +1,346 @@
+		},
+		Status: corev1.NamespaceStatus{},
+	}
+	client.CoreV1().Namespaces().Create(ctx, nsObj, metav1.CreateOptions{})
```
Error checking here and similar places would be helpful. I know that the check for eventual appearance comes later, but if something goes wrong then reporting the error would be helpful.

effi-ofer replied:
done

MikeSpreitzer replied:
Not seeing it in commit 73515206ca859c31aebbf5d56b010dc5e1cfdea7 .

<img width="667" alt="Screenshot 2024-03-11 at 11 22 17 AM" src="https://github.com/kubestellar/kubestellar/assets/14296719/73a8d2b7-cd53-4485-83de-9dbdab73d65b">

effi-ofer replied:
I took it out.  We often use the tests to setup the environment and then either run a subset of the tests or run something manually.  If the environment is already setup then some of the tests will fail when trying to re-create the namespace.

MikeSpreitzer replied:
The k8s errors package has a func to distinguish the error that means "already exists".

effi-ofer replied:
Good point.  I added a check for already exists.

effi-ofer commented:
These tests have so far uncovered these issues: #1710, #1711, #1838, #1850, #1868.

MikeSpreitzer commented on test/e2e/ginkgo/multiple_cluster_deployment_test.go:
```diff
@@ -0,0 +1,305 @@
+			util.ValidateNumDeployments(ctx, wec2, ns, 0)
+		})
+
+		ginkgo.It("deletes WEC objects when BindingPolicy is deleted", func() {
+			err := ksWds.ControlV1alpha1().BindingPolicies().Delete(ctx, "nginx", metav1.DeleteOptions{})
+			gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
+			util.ValidateNumDeployments(ctx, wec1, ns, 0)
+			util.ValidateNumDeployments(ctx, wec2, ns, 0)
+		})
```
Isn't the same test above in lines 86--92?

effi-ofer replied:
Yes. Good catch. Don't know how it snack in there.

MikeSpreitzer replied:
The duplication has not been removed yet.

effi-ofer replied:
fixed.

MikeSpreitzer commented on test/e2e/ginkgo/multiple_cluster_deployment_test.go:
```diff
@@ -0,0 +1,305 @@
+			util.ValidateNumDeployments(ctx, wec2, ns, 1)
+		})
+
+		/***
```
FYI: I suggest `if false {...}` rather than commenting out. That way we are forced to keep the code compiling.

effi-ofer replied:
done

MikeSpreitzer commented on test/e2e/ginkgo/multiple_cluster_deployment_test.go:
```diff
@@ -0,0 +1,305 @@
+		})
+		***/
+
+		ginkgo.It("downsync based on label and object name", func() {
```
label on what? DownsyncObjectTest can test labels on objects and also can test labels on the objet's Namespace.

effi-ofer replied:
changed to object labels.

MikeSpreitzer commented on test/util/util.go:
```diff
@@ -0,0 +1,346 @@
+			Namespace: "",
+			Labels:    labels,
+		},
+		Status: corev1.NamespaceStatus{},
```
FYI: I do not think you need this line.
See https://github.com/kubernetes/api/blob/v0.28.2/core/v1/types.go#L5831

effi-ofer replied:
removed.

MikeSpreitzer commented on test/e2e/README.md:
```diff
@@ -1,5 +1,6 @@
+- [all the end to end tests](ginkgo)
+- [multi-cluster deployment using shell scripts](multi-cluster-deployment)
+- [singleton status return using shell scripts](singleton-status)
+- [object fields filtering using shell scripts](object-cleaning)
```
Looks like a rebase lost a change here. It should be "object cleaning" rather than "object fields filtering".

effi-ofer replied:
fixed.

MikeSpreitzer commented on test/e2e/ginkgo/multiple_cluster_deployment_test.go:
```diff
@@ -0,0 +1,294 @@
+			util.ValidateNumDeployments(ctx, wec2, ns, 1)
+		})
+
+		ginkgo.It("handles OR of cluster and object selectors", func() {
```
This is not an accurate description of what is being tested for the cluster label selectors.

effi-ofer replied:
changed to "handles clusterSelector with MatchLabels and MatchExpressions"

MikeSpreitzer commented on test/e2e/ginkgo/multiple_cluster_deployment_test.go:
```diff
@@ -0,0 +1,294 @@
+				[]metav1.LabelSelector{
+					{MatchLabels: map[string]string{"name": "cluster1"}},
+					{MatchExpressions: []metav1.LabelSelectorRequirement{
+						{Key: "name", Operator: metav1.LabelSelectorOpIn, Values: []string{"cluster2"}},
```
I have seen code in the Kubernetes libraries for selectors that simplifies `in <set with one member>` to `== <that one member>`. So I would test with a set with multiple members.

effi-ofer replied:
Changed to 
```
				[]metav1.LabelSelector{
					{MatchLabels: map[string]string{"name": "cluster1"}},
					{MatchExpressions: []metav1.LabelSelectorRequirement{
						{Key: "name", Operator: metav1.LabelSelectorOpIn, Values: []string{"cluster1", "cluster2", "cluster3"}},
					}},
```

MikeSpreitzer reviewed (commented):
There is still an open question about KUBECONFIG in the GitHub workflow. But this is good enough to merge.

/lgtm
Thanks!

ezrasilvera commented:
/approve
